CMakeLists.txt

```diff
@@ -119,6 +120,12 @@ if(USE_HDFS)
     SET(HDFS_CXX_LIBRARIES ${HDFS_LIB} ${JAVA_JVM_LIBRARY})
 endif(USE_HDFS)
 
+if(USE_BOOST_THREAD)
+    message(STATUS "Using boost threading library")
+    ADD_DEFINITIONS(-DUSE_BOOST_THREAD)
+    find_package(Boost COMPONENTS thread REQUIRED)
```

Comment:
If `Boost` is required, I prefer `c++-17`.  Except for GPU version, we always try to reduce the external dependencies as possible, for better cross-platform compatibility, and avoid the merge issues or possible bugs from upstream.

Reply:
I just wanted to test if it worked. I will revert.

I just think that implementing it ourselves or trusting someone's implementation that is not standard can open a box of problems bigger than the ones that mantaining c++17 or extra dependencies would mean

Reply:
Discovering a bug in the custom implementation would be much more difficult to deal with in my opinion I mean